modules/ding_innskraning/ding_innskraning.module

```diff
@@ -0,0 +1,780 @@
+  }
+  else {
+    // OK, let's log the user in.
+    $login_success = _ding_innskraning_activate_user_by_ssn($ssn);
+  }
+
+  if ($login_success) {
+    // Go to frontpage.
+    drupal_goto('<front>');
+  }
+  else {
+    // Go to login page.
+    drupal_set_message(
+      t('Authentication through @link failed! Authentication through @link only available to registered users.',
+        array('@link' => l(t('island.is'), 'https://island.is'))
+      ),
+      'error'
+    );
+    drupal_goto('user/login');
+  }
+}
+
+/**
+ * Generate a SAML auth id.
+ *
+ * Creates an id for the authentication request. island.is requires the id to be
+ * formatted as a GUID.
+ */
+function _ding_innskraning_auth_id() {
+  // Implementation created from
+  // https://www.texelate.co.uk/blog/create-a-guid-with-php
+  // Create a token.
+  $token = $_SERVER['HTTP_HOST'];
+  $token .= $_SERVER['REQUEST_URI'];
+  $token .= uniqid(rand(), TRUE);
+
+  // GUID is 128-bit hex.
+  $hash = strtoupper(md5($token));
+
+  // Create formatted GUID.
+  $guid = '';
+
+  // GUID format is XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX for readability.
+  $guid .= substr($hash, 0, 8) . '-' . substr($hash, 8, 4) . '-' . substr($hash, 12, 4) .
+    '-' . substr($hash, 16, 4) . '-' . substr($hash, 20, 12);
+
+  return $guid;
+}
+
+/**
+ * Create the login request with SAML parameters.
+ */
+function _ding_innskraning_login_request($id, $issuer, $destination) {
+  // Add the destination URL for login.
+  $destination = url($destination, array('query' => array('id' => $issuer, 'auth' => $id)));
+  $destination = htmlspecialchars($destination);
+  // Take the name of the site as provider name.
+  $provider = variable_get('site_name', '');
+  // Log the time when request was created.
+  $issueInstant = format_date(time(), 'custom', "Y-m-dTH:i:sZ");
+  $providerNameStr = '';
+  // Consumer URL.
+  $consume_url = url('saml/consume', array('absolute' => TRUE));
+
+  if ($provider != '') {
+    $providerNameStr = <<<PROVIDERNAME
+ProviderName="{$provider}"
+PROVIDERNAME;
+  }
+
+  $request = <<<AUTHNREQUEST
+<samlp:AuthnRequest
+  xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
+  xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
+  ID="$id"
+  Version="2.0"
+  {$providerNameStr}
+  IssueInstant="$issueInstant"
+  Destination="{$destination}"
+  ProtocolBinding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"
+  AssertionConsumerServiceURL="{$consume_url}">
+  <saml:Issuer>{$issuer}</saml:Issuer>
+    <samlp:NameIDPolicy
+        Format="urn:oasis:names:tc:SAML:2.0:nameid-format:transient"
+        AllowCreate="true" />
+    <samlp:RequestedAuthnContext Comparison="exact">
+        <saml:AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:X509</saml:AuthnContextClassRef>
+    </samlp:RequestedAuthnContext>
+</samlp:AuthnRequest>
+AUTHNREQUEST;
+
+  $deflatedRequest = gzdeflate($request);
+  $base64Request = base64_encode($deflatedRequest);
+
+  return $base64Request;
```

Comment:
This incantation already seems to be handled by [`AuthnRequest` from `php-saml`](https://github.com/onelogin/php-saml/blob/master/lib/Saml2/AuthnRequest.php). Why is this not used?

Reply:
I do not know. That was how the module was implemented, and I was tasked with getting it to work, not re-implement it.

There might be a reason for not using AuthnRequest or there might not be, hard to say without a history of the module.